src/netconf_confirmed_commit.c

```diff
@@ -0,0 +1,938 @@
+ * @return SR_ERR_OK On success
+ */
+static int
+backup_module(sr_session_ctx_t *session, struct lys_module *module)
+{
+    int rc = SR_ERR_OK;
+    char *path = NULL, *xpath = NULL;
+    struct lyd_node *node = NULL;
+
+    /* LOCK commit_ctk.lock */
+    pthread_mutex_lock(&commit_ctx.lock);
+
+    VRB("Backing up module: %s", module->name);
+
+    if (asprintf(&xpath, "/%s:*", module->name) == -1) {
+        rc = SR_ERR_NO_MEMORY;
+        goto cleanup;
+    }
+
+    if ((rc = sr_get_data(session, xpath, 0, 0, SR_OPER_NO_STATE, &node))) {
+        ERR("Failed getting configuration of running for module \"%s\".", module->name);
+        goto cleanup;
+    }
+
+    if ((rc = check_dir())) {
+        goto cleanup;
+    }
+    if (asprintf(&path, "%s/%s.json", NP2SRV_DIR, module->name) == -1) {
+        rc = SR_ERR_NO_MEMORY;
+        goto cleanup;
+    }
+    if (lyd_print_path(path, node, LYD_JSON, LY_PRINT_SHRINK)) {
+        rc = SR_ERR_LY;
+        goto cleanup;
+    }
+
+cleanup:
+    lyd_free_tree(node);
+    free(path);
+    free(xpath);
+    /* UNLOCK commit_ctk.lock */
+    pthread_mutex_unlock(&commit_ctx.lock);
+    return rc;
+}
+
+/**
+ * @brief Create the timer and set all the options.
+ *
+ * @param[in] timeout Time after which the timer will start the thread.
+ *
+ * @return SR_ERR_SYS When creating the timer fails.
+ * @return SR_ERR_OK When succeeded.
+ */
+static int
+start_timer_thread(uint32_t timeout)
+{
+    struct sigevent sev;
+    struct itimerspec its;
+    timer_t timer_id;
+
+    /* reset the structures */
+    memset(&sev, 0, sizeof sev);
+    memset(&its, 0, sizeof its);
+
+    /* create and arm the timer */
+    sev.sigev_notify = SIGEV_THREAD;
+    sev.sigev_notify_function = changes_rollback;
+    its.it_value.tv_sec = timeout;
+    if (timer_create(CLOCK_REALTIME, &sev, &timer_id) == -1) {
+        return SR_ERR_SYS;
+    }
+    if (timer_settime(timer_id, 0, &its, NULL) == -1) {
+        return SR_ERR_SYS;
+    }
+    set_timer(timer_id);
+
+    return SR_ERR_OK;
+}
+
+/**
+ * @brief Get information from restore file and remove it.
+ *
+ * @param[out] time When the confirmed commit was called.
+ * @param[out] timeout What was the timeout supposed to be.
+ * @return SR_ERR_SYS When reading file failed;
+ * @return SR_ERR_OK When successful. If time is zero no restore file existed.
+ */
+static int
+read_restore_file(time_t *time, uint32_t *timeout)
+{
+    FILE *file;
+
+    *time = 0;
+    *timeout = 0;
+
+    /* Check for file */
+    if (access(NP2SRV_DIR RESTORE_FILE, F_OK)) {
+        return SR_ERR_OK;
+    }
+
+    file = fopen(NP2SRV_DIR RESTORE_FILE, "r");
+    if (!file) {
+        ERR("Confirmed commit restore file found but not readable");
+        return SR_ERR_SYS;
+    }
+
+    if (fscanf(file, "%ld\n%d\n", time, timeout) != 2) {
+        ERR("Malformed confirmed commit restore file. Could not recover.");
+        *time = 0;
+        *timeout = 0;
+        fclose(file);
```

Comment:
File is not closed on success. Probably a good candidate for having a proper cleanup where you can also zero `time` and `timeout` on error.